plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,217 @@
+  enum { DATA = 0, LABEL = 1 };
+
+  virtual bool Next(void) {
+    if (caffe_data_layer_) {
+      // MxNet iterator is expected to return CPU-accessible memory
+      assert(::caffe::Caffe::mode() == ::caffe::Caffe::CPU);
+      caffe_data_layer_->Forward(bottom_, top_);
+      assert(batch_size_ > 0);
+
+      if (loc_ + batch_size_ <= param_.max_iterations) {
+        batch_data_.dptr_ = top_[DATA]->mutable_cpu_data();
+        batch_label_.dptr_ = top_[LABEL]->mutable_cpu_data();
+
+        out_.data.clear();
+        if (param_.flat) {
+          out_.data.push_back(TBlob(batch_data_.FlatTo2D()));
+        } else {
+          out_.data.push_back(TBlob(batch_data_));
+        }
+        assert(out_.batch_size == batch_size_);
+        out_.data.push_back(TBlob(batch_label_));
+        loc_ += batch_size_;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  virtual const TBlobBatch &Value(void) const {
+    return out_;
+  }
+
+ private:
+  /*! \brief MNISTCass iter params */
+  CaffeDataParam param_;
+  /*! Shape scalar values */
+  size_t batch_size_, channels_, width_, height_;
+  /*! \brief Caffe data layer */
+  boost::shared_ptr<caffe::Layer<Dtype> >  caffe_data_layer_;
+  /*! \brief batch data tensor */
+  mshadow::Tensor<cpu, 4, Dtype> batch_data_;
+  /*! \brief batch label tensor  */
+  mshadow::Tensor<cpu, 2, Dtype> batch_label_;
+  /*! \brief Output blob data for this iteration */
+  TBlobBatch out_;
+  /*! \brief Bottom and top connection-point blob data */
+  std::vector<::caffe::Blob<Dtype>*> bottom_, top_;
+  /*! \brief Cleanup these blobs on exit */
+  std::list<std::unique_ptr<::caffe::Blob<Dtype>>> cleanup_blobs_;
+  /*! \brief Blobs done so far */
+  std::atomic<size_t>  loc_;
+};  // class CaffeDataIter
+
+class CaffeDataPrefetcherIter : public PrefetcherIter {
+ public:
+    CaffeDataPrefetcherIter() : PrefetcherIter(NULL) {}
+    virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+      param_.InitAllowUnknown(kwargs);
+      switch (param_.dtype) {
+        case mshadow::kFloat32:
+          this->loader_.reset(new CaffeDataIter<float>());
+          break;
+        case mshadow::kFloat64:
+          this->loader_.reset(new CaffeDataIter<double>());
+          break;
+        case mshadow::kFloat16:
+          LOG(FATAL) << "float16 layer is not supported by caffe";
+          return;
+        default:
+          LOG(FATAL) << "Unsupported type " << param_.dtype;
+          return;
+      }
+      PrefetcherIter::Init(kwargs);
+    }
+
+ protected:
+    CaffePrefetchParam param_;
+};
+
+DMLC_REGISTER_PARAMETER(CaffeDataParam);
+DMLC_REGISTER_PARAMETER(CaffePrefetchParam);
+
+MXNET_REGISTER_IO_ITER(CaffeDataIter)
+.describe("Create MxNet iterator for a Caffe data layer.")
+.add_arguments(CaffeDataParam::__FIELDS__())
+.add_arguments(PrefetcherParam::__FIELDS__())
+.add_arguments(CaffePrefetchParam::__FIELDS__())
+.set_body([]() {
+    return new CaffeDataPrefetcherIter();
```

Comment:
Yeah, I saw that -- I was not sure if all Caffe data layers tend to derive from BasePrefetchingDataLayer or not. I suppose I can have a parameter determine if prefetching is desired? Or you think I should just treat all Caffe data layers as prefetching data layers?